src/libsync/ocsprofileconnector.cpp

```diff
@@ -0,0 +1,148 @@
+        }
+        QImage scaledSvg(imageSize, QImage::Format_ARGB32);
+        scaledSvg.fill("transparent");
+        QPainter svgPainter{&scaledSvg};
+        svgRenderer.render(&svgPainter);
+        return QPixmap::fromImage(scaledSvg);
+    }
+    return {};
+}
+
+OCC::Optional<QPixmap> iconDataToPixmap(const QByteArray iconData)
+{
+    if (!iconData.startsWith("<svg")) {
+        return {};
+    }
+    return createPixmapFromSvgData(iconData);
+}
+}
+
+namespace OCC {
+
+HovercardAction::HovercardAction() = default;
+
+HovercardAction::HovercardAction(QString title, QUrl iconUrl, QUrl link)
+    : _title(std::move(title))
+    , _iconUrl(std::move(iconUrl))
+    , _link(std::move(link))
+{
+}
+
+OcsProfileConnector::OcsProfileConnector(AccountPtr account, QObject *parent)
+    : QObject(parent)
+    , _account(account)
+{
+}
+
+void OcsProfileConnector::fetchHovercard(const QString &userId)
+{
+    const QString url = QStringLiteral("/ocs/v2.php/hovercard/v1/%1").arg(userId);
+    auto job = new JsonApiJob(_account, url, this);
+    connect(job, &JsonApiJob::jsonReceived, this, &OcsProfileConnector::onHovercardFetched);
+    job->start();
+}
+
+void OcsProfileConnector::onHovercardFetched(const QJsonDocument &json, int statusCode)
+{
+    qCDebug(lcOcsProfileConnector) << "Hovercard fetched:" << json;
+
+    if (statusCode != 200) {
+        qCInfo(lcOcsProfileConnector) << "Fetching of hovercard finished with status code" << statusCode;
+        return;
+    }
+    const auto jsonData = json.object().value("ocs").toObject().value("data").toObject().value("actions");
+    Q_ASSERT(jsonData.isArray());
+    _currentHovercard = jsonToHovercard(jsonData.toArray());
+    fetchIcons();
+    emit hovercardFetched();
+}
+
+void OcsProfileConnector::setHovercardActionIcon(const std::size_t index, const QPixmap &pixmap)
+{
+    auto &hovercardAction = _currentHovercard._actions[index];
+    hovercardAction._icon = pixmap;
+    emit iconLoaded(index);
+}
+
+void OcsProfileConnector::loadHovercardActionIcon(const std::size_t hovercardActionIndex, const QByteArray &iconData)
+{
+    if (hovercardActionIndex >= _currentHovercard._actions.size()) {
+        // Note: Probably could do more checking, like checking if the url is still the same.
+        return;
+    }
+    const auto icon = iconDataToPixmap(iconData);
+    if (icon.isValid()) {
+        setHovercardActionIcon(hovercardActionIndex, icon.get());
+        return;
+    }
+    qCWarning(lcOcsProfileConnector) << "Could not load Svg icon from data" << iconData;
+}
+
+void OcsProfileConnector::startFetchIconJob(const std::size_t hovercardActionIndex)
+{
+    const auto hovercardAction = _currentHovercard._actions[hovercardActionIndex];
+    const auto iconJob = new IconJob{hovercardAction._iconUrl, this};
+    connect(iconJob, &IconJob::jobFinished, [this, hovercardActionIndex](const QByteArray &iconData) {
+        loadHovercardActionIcon(hovercardActionIndex, iconData);
```

Comment:
so what happens here in case the `IconJob` fails ?

Reply:
Good point, I need to add an API for that inside `IconJob`.

Reply:
But, what should we do in case of an error? Can we not ignore the error and just don't show any image? Should we propagate the error to the user?

Reply:
can't we use a placeholder image and just add the error to the logs?

Reply:
We can log the error, but I don't see any benefit in adding a placeholder image. We still have the text anyway. Not sure a placeholder image will improve the user experience.